File: Lab3/Task1.m
B = 10;
A = 0.01;
fs = 5*B;

t_start=-10;
t_end= 10;
t = t_start:1/fs:t_end;

h_t = 2*0.1*sinc(2*B*t);

N = length(h_t);
h_f = fft(h_t, N)/fs;
h_f_abs = abs(h_f);
freq_axis = linspace(-fs/2, fs/2, N);

figure
subplot(2,1,1); 
plot(t, h_t)
title('Channel $h(t)$', 'interpreter', 'latex');
xlabel('Time(s)', 'interpreter', 'latex')
ylabel('Amplitude', 'interpreter', 'latex')
legend('$h(t)$', 'interpreter', 'latex');
grid on;

subplot(2,1,2); 
plot(freq_axis,fftshift(h_f_abs))
grid on;
xlabel('Frequency(Hz)', 'interpreter', 'latex');
ylabel('Amplitude', 'interpreter', 'latex');
legend('$H(f)$', 'interpreter', 'latex');

File: Lab4/Task1.py
import numpy as np
import matplotlib.pyplot as plt
from numpy.fft import fft
import random
import time


start_time = 0
stop_time = 1
fm = 10  # Maximum frequency component in Hertz for the given spectrum
fs = 10 * fm
ts = 1 / fs
mx = np.array([])
tvect = np.array([])


for T in range(30):
    time = np.arange(start_time+T, stop_time+T, ts)
    U = random.randint(1,5)
    m_t = U*np.cos(2*fm*time)
    mx = np.append(mx, m_t)
    tvect = np.append(tvect, time)
    mf = fft(mx)/fs
    N = len(mf)
    mf_abs_sorted = np.fft.fftshift(abs(mf))
    freq_axis = np.linspace(-fs/2, fs/2, N)

    # Time Domain
    plt.figure(1)
    plt.plot(tvect, mx)
    plt.title(f'Time Domain, Time=({start_time+T}, {stop_time+T})')
    plt.xlabel('time')
    plt.ylabel('Amplitude')

    # Save the final figure
    if T==29:
        plt.savefig('task1_final_time.svg')

    # Frequency Domain
    plt.figure(2)
    plt.plot(freq_axis, mf_abs_sorted)
    plt.xlabel('Frequency(Hz)')
    plt.ylabel('Amplitude')
    plt.title(f'Frequency Domain, Time=({start_time+T}, {stop_time+T})')

    # Save the final figure
    if T==29:
        plt.savefig('task1_final_fft.svg')

    plt.show()
    plt.pause(0.5)